feat(core): support encapsulated nx install (#14705)
nrwl/nx: docs/shared/guides/encapsulated-nx-and-the-wrapper.md

# Encapsulated Nx

> `nx init` accepts a flag called `--encapsulated`. When this flag has been passed, Nx manages its own installation separately from any dependencies that are local to your repository. As such, `nx` and any plugins will not be listed in a root level package.json, and also will not be in a root level node_modules folder. In fact, these are not required for an encapsulated Nx repository.

Reviewer: ```suggestion
`nx init` accepts a flag called `--encapsulated`. When this flag has been passed, Nx manages its own installation within a `.nx` directory.
```

Support for `--encapsulated` was added in Nx v15.8.0. To ensure that it is available, specify the version of nx when running your command so that `npx` doesn't accept an older version that is in the cache. (e.g. `npx nx@latest init --encapsulated` or `npx nx@15.8.0 init --encapsulated`)

{% /callout %}

> `nx init` accepts a flag called `--encapsulated`. When this flag has been passed, Nx manages its own installation within a `.nx` directory.

Reviewer 2: ```suggestion
`nx init` accepts a flag called `--encapsulated`. When this flag is passed, Nx manages its own installation within a `.nx` directory.  This allows you to use Nx without a `node_modules` folder.
```

> `nx init` accepts a flag called `--encapsulated`. When this flag has been passed, Nx manages its own installation within a `.nx` directory. This allows using Nx without a `node_modules` folder or root `package.json`.

Reviewer 2: ```suggestion
`nx init` accepts a flag called `--encapsulated`. When this flag is passed, Nx manages its own installation within a `.nx` directory. This allows Nx to be used without a `node_modules` folder or root `package.json`.
```

- [nx.json](#nx-json)

> {% callout type="note" title="Available since Nx v15.7.0" %}

Reviewer: ```suggestion
{% callout type="note" title="Available since Nx v15.8.0" %}
```

> {% callout type="note" title="Available since Nx v15.7.0" %}

Reviewer: Let's move this to the top of the doc.

> Support for `--encapsulated` was added in Nx v15.7.0. To ensure that it is available, specify the version of nx when running your command so that `npx` doesn't accept an older version that is in the cache. (e.g. `npx nx@15.7.0 init --encapsulated` or `npx nx@latest init --encapsulated`)

Reviewer: ```suggestion
Support for `--encapsulated` was added in Nx v15.8.0. To ensure that it is available, specify the version of nx when running your command so that `npx` doesn't accept an older version that is in the cache. (e.g. `npx nx@latest init --encapsulated` or `npx nx@15.8.0 init --encapsulated`)
```

- [nx.bat](#nx-and-nxbat)
- [nx.json](#nx-json)

## Nx Wrapper

> When `nx init --encapsulated` is ran, a file is created called `.nx/nxw.js`. This javascript file is responsible for ensuring that your local installation of Nx is up to date with what the repository expects, and then invoking Nx. We refer to this as the "Nx Wrapper".

Reviewer 2: ```suggestion
When `nx init --encapsulated` is run, a file is created called `.nx/nxw.js`. This javascript file is responsible for ensuring that your local installation of Nx is up to date with what the repository expects, and then invoking Nx. We refer to this as the "Nx Wrapper".
```

## Nx and nx bat

> `nx` and `nx.bat` perform the same function, with `nx` working under the bash shell and `nx.bat` working for windows users. Each script simply checks that `node` and `npm` are available, and then invokes the [nx wrapper](#nx-wrapper). They should each be committed to your repository, so that users of windows, linux, and macos are able to use Nx.

Reviewer: ```suggestion
`nx` and `nx.bat` perform the same function, with `nx` working under the bash shell and `nx.bat` working for Windows users. Each script simply checks that `node` and `npm` are available, and then invokes the [nx wrapper](#nx-wrapper). They should each be committed to your repository, so that developers on any operating system can use Nx.
```

To invoke an Nx command, you would use:

{% tabs %}
{% tab label="MacOS / Linux" %}

```shell
> ./nx build my-project
> ./nx generate application
> ./nx graph
```

{% /tab %}
{% tab label="Windows" %}

```shell
> ./nx.bat build my-project
> ./nx.bat generate application
> ./nx.bat graph
```

{% /tab %}
{% /tabs %}

## Nx Json

> This is the configuration file for Nx. A full description of `nx.json`, and the possible configuration options, is located [here](/reference/nx-json).

Reviewer 2: ```suggestion
This is the configuration file for Nx. A full description of `nx.json`, and the possible configuration options, is located in [the nx.json reference documentation](/reference/nx-json).
```

> When using Nx in encapsulated mode, an `installation` key is added to `nx.json`. This property tracks the currently installed version of Nx and any plugins.
> 
> A full description of `nx.json`, and the possible configuration options, is located [here](/reference/nx-json).

Reviewer: ```suggestion
This is the configuration file for Nx. A full description of `nx.json`, and the possible configuration options, is located [here](/reference/nx-json).

When Nx is encapsulated, an `installation` key is added to `nx.json`. This property tracks the currently installed version of Nx and plugins.
```
